fq.go

```diff
@@ -0,0 +1,163 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+)
+
+type FQScheduler struct {
+	lock   *sync.Mutex
+	queues []*Queue
+	vt     uint64
+	C      uint64
+	G      uint64
+	seen   bool
+}
+
+// TODO(aaron-prindle) add concurrency enforcement - 'C'
+
+func (q *FQScheduler) chooseQueue(packet *Packet) *Queue {
+	for _, queue := range q.queues {
+		if packet.key == queue.key {
+			return queue
+		}
+	}
+	panic("no matching queue for packet")
+}
+
+func newFQScheduler(queues []*Queue) *FQScheduler {
+	now := NowAsUnixMilli()
+	fq := &FQScheduler{
+		lock:   &sync.Mutex{},
+		queues: queues,
+		// R(t) = (server start time) + (1 ns) * (number of rounds since server start).
+		vt: now,
+	}
+	// TODO(aaron-prindle) verify if this is needed?
+	for i := range fq.queues {
+		fq.queues[i].virstart = fq.vt
+	}
+
+	return fq
+}
+
+// TODO(aaron-prindle) verify that the time units are correct/matching
+func NowAsUnixMilli() uint64 {
+	return uint64(time.Now().UnixNano() / 1e6)
+}
+
+func (q *FQScheduler) processround() (*Packet, bool) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.tick()
+	return q.Dequeue()
+}
+
+func (q *FQScheduler) Enqueue(packet *Packet) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	fmt.Printf("enqueue: %d\n", packet.key)
+	q.seen = true
+
+	queue := q.chooseQueue(packet)
+	// TODO(aaron-prindle) q.now() or get virstart of the queue you go in?
+	packet.starttime = q.now()
```

Comment:
It looks to me like `packet.starttime` is intended to be the real time when the packet started executing, not when it arrived to the queue.  See how later it is subtracted from the real finish time to get the real service duration?

What actually needs to done at this point is to set the queue's virtual start time to the current virtual time if the queue was empty and had no requests executing.

Reply:
Makes sense I will modify this.  It was not clear to me if the adjustment was in virtual time or real time.